### PR TITLE
CLI Improvements

### LINKED ISSUE #561: [TODO]: fix CLI and add CLI integration tests
**Notes**
Recent changes have broken the CLI (training and prediction for sure, and maybe others).

In addition to fixing the CLI, we should add integration tests that run the CLI from Python to catch when it breaks in the future.

### PR DESCRIPTION
**Pull request by KnathanM**
## Description
This PR works on #561 by fixing parts of the CLI that worked before and adding some CLI integration tests. This PR is to just get the basic CLI working again. Future PRs will add more features to the CLI and tests for them. 

## Future workflow
The CLI was last worked on in early October. There have been many changes to the code since then and new abilities added. We have focused on making the code work in a notebook setting but also want the code to be accessible from the command line. 

We want command line users to be able to access most if not all of the code features. We also want to make it easy for v1 users to switch their workflows to v2. As such we are keeping track of CLI changes in this spreadsheet: https://docs.google.com/spreadsheets/d/1PxfHrxqD_XmY4IfoTat0HnkJrGKKRTg_oyfalOVNNXg/edit#gid=0

## Checklist
- [x] linted with flake8?
- [x] (if appropriate) unit tests added?

### CODE REVIEW

File: chemprop/cli/train.py [hunk 1]
```diff
@@ -621,12 +621,13 @@ def main(args):
         results = trainer.test(model, test_loader)[0]
         logger.info(f"Test results: {results}")
 
-    p_model = args.output_dir / "model.pt"
+    p_model = args.output_dir / "model.ckpt"
     torch.save(model.state_dict(), p_model)
     logger.info(f"model state dict saved to '{p_model}'")
```
Comment (hwpang) [inline chemprop/cli/train.py:626]: I think during today's chemprop v2 dev meeting, we think this can be removed.

**Review by hwpang — DISMISSED** [review]
Thanks for the PR! I see the part that are intertwined with the multicomponent. Maybe I can push some changes directly to this PR draft to add the CLI for multicomponent. Additionally, we should add tests to test the CLI.

File: chemprop/cli/predict.py [hunk 2]
```diff
@@ -159,7 +158,6 @@ def validate_predict_args(args):
 def main(args):
     model = MPNN.load_from_checkpoint(args.checkpoint_path)
 
-    bond_messages = isinstance(model.message_passing, BondMessagePassing)
```
[the remark this answers is not in the document]
Comment (KnathanM) [reply chemprop/cli/predict.py:312]: bond_messages was passed through data set creation to be given to featurization. My best guess is it was never needed and accidently added by David in his large refactoring. I don't actually know what it practically means for messages to be centered on bonds or atoms, but in the v1 code this wasn't needed for featurization.

File: chemprop/cli/utils/parsing.py [hunk 3]
```diff
@@ -200,21 +200,15 @@ def build_data_from_files(
...
     if isinstance(data[0], MoleculeDatapoint):
         extra_atom_fdim = data[0].V_f.shape[1] if data[0].V_f is not None else 0
         extra_bond_fdim = data[0].E_f.shape[1] if data[0].E_f is not None else 0
-        featurizer = MoleculeMolGraphFeaturizer(
-            bond_messages=bond_messages,
-            extra_atom_fdim=extra_atom_fdim,
-            extra_bond_fdim=extra_bond_fdim,
+            extra_atom_fdim=extra_atom_fdim, extra_bond_fdim=extra_bond_fdim
         )
         return MoleculeDataset(data, featurizer)
```
[the remark this answers is not in the document]
Comment (KnathanM) [reply chemprop/cli/utils/parsing.py]: That is an artifact of how I split up my changes into commits. Line 209 should not have been removed here. My change was to change `MoleculeMolGraphFeaturizer` to `SimpleMoleculeMolGraphFeaturizer`, which happens in the 5th commit.

File: pyproject.toml [hunk 4]
```diff
@@ -33,6 +33,7 @@ dependencies = [
     "rdkit",
     "scikit-learn",
     "scipy",
+    "tensorboard",
```
Comment (KnathanM) [inline pyproject.toml:36]: tensorboard might already be an implicit dependency through pytorch lightning. My conda environment already had `tensorboardx=2.6.2` without my specification. This [file](https://github.com/Lightning-AI/pytorch-lightning/blob/93c1ab0653d50db1fb6c3cd3a7c26f4cee5cecac/requirements/pytorch/test.txt#L18) may be related.

Comment (hwpang) [reply pyproject.toml:36]: That's weird, because the github workflow fails due to the dependency issue [here](https://github.com/chemprop/chemprop/actions/runs/7574807468/job/20630020134#step:6:28) when testing the CLI part, and that's why I added it as a dependency explicitly.

**Comment by KnathanM** [comment]
The basic CLI works again and Hao-Wei has added some tests. This could be enough to change the scope of the PR and merge in. Thoughts?

**Comment by hwpang** [comment]
I see the test failed [here](https://github.com/chemprop/chemprop/actions/runs/7576938285/job/20636771994?pr=585#step:6:25), which seems to be a file name incompatibility issue in windows?

**Comment by KnathanM** [comment]
> I see the test failed [here](https://github.com/chemprop/chemprop/actions/runs/7576938285/job/20636771994?pr=585#step:6:25), which seems to be a file name incompatibility issue in windows?

I see that too, but I don't see what the error is. If you have any ideas, great. Or I'm planning to look at it again later tonight.

**Comment by JacksonBurns** [comment]
@KnathanM @hwpang Windows does not allow the colon character (i.e. `:`) in filenames. You can still use a timestamp in the logfile name, but you need to remove or replace the colons.

**Comment by hwpang** [comment]
@JacksonBurns Thanks for the tip! @KnathanM and I pushed commits to fix this by replacing the colons with dashes.

File: pyproject.toml (hunk 4 above)
Comment (KnathanM) [reply pyproject.toml:36]: @JacksonBurns Any idea why we need to be explicit with the tensorboard dependency for the testing while when I installed from conda, it got automatically installed with lightning?

File: chemprop/cli/predict.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidegraff) [reply chemprop/cli/predict.py:312]: per slack, this can be removed. this was needed in v2-alpha back when bond- vs. atom-message passing was determined by graph featurization. It's now handled inside the message passing operation.

File: chemprop/cli/train.py (hunk 1 above)
Comment (davidegraff) [reply chemprop/cli/train.py]: There's a bit of an over-reliance on `load_from_checkpoint`, a bit like using a sledgehammer to crack a nut. It solves the problem for loading model state to make predictions, but it muddies the water between a "checkpoint" and "model state." The former contains all the information to continue training, and the latter contains just the weight. Of course, you can't just load weights without knowing the model's architecture, so we need to load that elegantly as well, at which point you might as well just couple them together when saving because one is fairly useless without the other.

To summarize, we should think about loading "model state + arch" when testing rather than assuming we have a checkpoint and just relying on `load_from_checkpoint` to do the trick. It might help first to copy code from `load_from_checkpoint` into a separate `@classmethod`, `load_model`, that loads both model architecture ("hyperparamters") and weights. It should be assumed that a "checkpoint" is a superset of a "model state" because, again, the former contains a model's state "and then some."

File: pyproject.toml (hunk 4 above)
Comment (davidegraff) [reply pyproject.toml:36]: we don't need it for the base package but do require it for the CLI. So your new CI is failing because you're trying to run the CI during testing.

File: .github/workflows/tests.yml [hunk 5]
```diff
@@ -40,6 +40,11 @@ jobs:
         shell: bash -l {0}
         run: |
           pytest -v tests
+      - name: Test CLI
+        shell: bash -l {0}
+        run: |
+          chemprop train -i tests/data/regression/mol.csv --task-type regression --epochs 1 --num-workers 4
+          chemprop predict --checkpoint tests/data/example_model_v2.ckpt -i tests/data/regression/mol.csv --num-workers 4
```
Comment (davidegraff) [inline .github/workflows/tests.yml:47]: should be a separate PR and should be inside of `pytest`

File: chemprop/cli/predict.py [hunk 6]
```diff
@@ -290,10 +139,10 @@ def add_predict_args(parser: ArgumentParser) -> ArgumentParser:
 
 
 def process_predict_args(args: Namespace) -> Namespace:
-    args.separate_test_path = Path(args.separate_test_path)
     if args.output is None:
-        name = f"{args.separate_test_path.stem}_preds.csv"
-        args.output = Path(args.separate_test_path.with_name(name))
+        args.test_path = Path(args.test_path)
+        name = f"{args.test_path.stem}_preds.csv"
+        args.output = Path(args.test_path.with_name(name))
```
Comment (davidegraff) [inline chemprop/cli/predict.py:145]:
```suggestion
        args.output =args.test_path.with_name(name)
```

File: pyproject.toml (hunk 4 above)
Comment (JacksonBurns) [reply pyproject.toml:36]: @KnathanM lightning used to ship with tensorboard (like, when you made your local environment), but they made it optional (so the CI that runs today won't include it). See this discussion if you're curious for more: https://github.com/Lightning-AI/pytorch-lightning/issues/9900

File: chemprop/cli/train.py (hunk 1 above)
Comment (hwpang) [reply chemprop/cli/train.py]: Thanks for the comments, @KnathanM and @davidegraff. I think it's best to add the `load_model` functionality in a separate PR. I will open up an issue to keep track of this.

File: .github/workflows/tests.yml (hunk 5 above)
Comment (hwpang) [reply .github/workflows/tests.yml:47]: Thanks! I will make an issue to keep track of this.

File: pyproject.toml (hunk 4 above)
Comment (KnathanM) [reply pyproject.toml:36]: That is exactly what I wanted to know. Thank @JacksonBurns! 
If lightning made tensorboard optional and defaults to CSVLogger when tensorboard isn't available, do we want to do the same? This would involve checking if the environment has tensorboard, which shouldn't be too difficult and would allow us to remove the tensorboard dependency.

**Review by davidegraff — APPROVED** [review]
